### PR TITLE
Test/jax core parameter client unit tests

### LINKED ISSUE #547: [TEST] Jax System Core Parameter Client
### What do you want to test?
Jax system core parameter client (in `mava/systems/jax/`)

### Outline of test structure
* Unit tests
* Test components and hooks

### Definition of done
Passing checks, cover all hooks, edge cases considered

### Mandatory checklist before making a PR
* [ ] The success criteria laid down in “Definition of done” are met.
* [ ] Test code is documented - docstrings for methods and classes, static types for arguments.
* [ ] Documentation is updated - README, CONTRIBUTING, or other documentation.

### PR DESCRIPTION
**Pull request by RuanJohn**
## What?
Unit tests for core parameter client
## Why?
Ongoing efforts to increase testing code coverage for callback redesign. 
## Extra
closes #547

### CODE REVIEW

File: tests/jax/core_system_components/parameter_client_test.py [hunk 1]
```diff
@@ -15,4 +15,406 @@
...
+    }
+
+    assert parameter_client._set_get_call_counter == 0
+
+    parameter_client.set_and_get_async()
+
+    assert parameter_client._set_get_future is None
+
+
+def test_add_async(parameter_client: ParameterClient) -> None:
```
Comment (mmorris44) [inline tests/jax/core_system_components/parameter_client_test.py:317]: I can't find a test here for what happens when `self._add_future is None`

File: tests/jax/core_system_components/parameter_client_test.py [hunk 2]
```diff
@@ -15,4 +15,406 @@
...
+
+    parameter_client.add_async(params={"new_key_1": 1})
+
+    assert parameter_client._async_add_buffer == {"new_key_1": 2}
+
+    # set add_future to dummy variable other than none
+    # assert that new parameter is added to async buffer
+    parameter_client._add_future._state = "not_done"  # type: ignore
+    parameter_client._async_add_buffer = {}
+    parameter_client.add_async(params={"new_key_2": 1})
```
Comment (mmorris44) [inline tests/jax/core_system_components/parameter_client_test.py:336]: This seems to test again what was just tested

File: tests/jax/core_system_components/parameter_client_test.py [hunk 3]
```diff
@@ -15,4 +15,406 @@
...
+                "networks-network_key_0": {
+                    "layer_0": {"weights": "new_weights", "biases": "new_biases"}
+                }
+            },
+        )
+
+
+def test_copy_array_with_device(parameter_client: ParameterClient) -> None:
+    """Test that new parameters are set on a device when a device \
+        is give."""
```
Comment (mmorris44) [inline tests/jax/core_system_components/parameter_client_test.py:388]: given

**Review by mmorris44 — CHANGES_REQUESTED** [review]
Very thorough and well-written PR, thanks @RuanJohn!
Just a few minor comments.

File: tests/jax/core_system_components/parameter_client_test.py [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # python3
-# Copyright 2021 InstaDeep Ltd. All rights reserved.
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply tests/jax/core_system_components/parameter_client_test.py:2]: Sorted.

File: tests/jax/core_system_components/parameter_client_test.py (hunk 1 above)
Comment (RuanJohn) [reply tests/jax/core_system_components/parameter_client_test.py:317]: By default `self._add_future` is set to `None`, please see [here](https://github.com/instadeepai/Mava/blob/9a0835a4998d17cab580a7269443c12cc16cf6d1/mava/systems/jax/parameter_client.py#L87). So we technically check for that condition in lines 319 - 322. I have also added an extra assert to verify that `self._add_future.done()` is `True` after adding the parameters. I can also explicitly set the value of `self.add_future` to `None` in the test if you think that would make it more explicit?

Comment (mmorris44) [reply tests/jax/core_system_components/parameter_client_test.py:317]: Ah I see, no worries then!
Maybe just add an assert before calling `add_async` that `self._add_future is None`, to ensure that the default is set correctly?

File: tests/jax/core_system_components/parameter_client_test.py (hunk 2 above)
Comment (mmorris44) [reply tests/jax/core_system_components/parameter_client_test.py:337]: Keep both of them. Your explanation makes sense

File: tests/jax/core_system_components/parameter_client_test.py (hunk 1 above)
Comment (RuanJohn) [reply tests/jax/core_system_components/parameter_client_test.py:317]: Added it :call_me_hand:

File: tests/jax/core_system_components/parameter_client_test.py [hunk 5]
```diff
@@ -15,4 +15,410 @@
...
+import jax
+import numpy as np
+import pytest
+
+from mava.callbacks.base import Callback
+from mava.systems.jax.parameter_client import ParameterClient
+from mava.systems.jax.parameter_server import ParameterServer
+
+
+class MockParameterServer(ParameterServer):
```
Comment (sash-a) [inline tests/jax/core_system_components/parameter_client_test.py:31]: Too late now, but it's for things like this where I wonder if we shouldn't have used monkeypatch

File: tests/jax/core_system_components/parameter_client_test.py [hunk 6]
```diff
@@ -15,4 +15,410 @@
...
+    assert parameter_client._parameters == {
+        "key_0": np.array(1, dtype=np.int32),
+        "key_1": np.array(2, dtype=np.float32),
+        "key_2": np.array(3, dtype=np.int32),
+        "key_3": np.array(4, dtype=np.int32),
+        "key_4": np.array(5, dtype=np.int32),
+        "networks-network_key_0": {"layer_0": {"weights": 1, "biases": 1}},
+        "networks-network_key_1": {"layer_0": {"weights": 2, "biases": 2}},
+        "networks-network_key_2": {"layer_0": {"weights": 3, "biases": 3}},
+        "allkey_0": np.array(0, dtype=np.int32),
```
Comment (sash-a) [inline tests/jax/core_system_components/parameter_client_test.py:307]: You expect this 3 times maybe extract into `expected_get_params`? But I can see an argument to leave it as is too.

**Review by sash-a — APPROVED** [review]
Super clear tests @RuanJohn, was a very easy review! Only 1 minor change, but I don't think it's strictly necessary so you can leave it as is.

File: tests/jax/core_system_components/parameter_client_test.py (hunk 6 above)
Comment (RuanJohn) [reply tests/jax/core_system_components/parameter_client_test.py:307]: I think it is a bit easier to leave it as is for now.

**Review by sash-a — APPROVED** [review]
:rocket:

**Review by mmorris44 — APPROVED** [review]
Re-approved. Still happy! Thanks @RuanJohn
